ros/angel_system_nodes/angel_system_nodes/uho_activity_detector.py

```diff
@@ -0,0 +1,341 @@
+
+        self.subscription_list = []
+        # Image subscription
+        self.subscription_list.append(mf.Subscriber(self, Image, self._image_topic))
+        # Hand pose subscription
+        self.subscription_list.append(mf.Subscriber(self, HandJointPosesUpdate, self._hand_topic))
+        self.time_sync = mf.ApproximateTimeSynchronizer(
+            self.subscription_list,
+            1500, # TODO: queue size
+            10  # TODO: slop (delay msgs can be synchronized in seconds)
+        )
+
+        self.time_sync.registerCallback(self.multimodal_listener_callback)
+
+        # Object detections subscriber
+        self._obj_det_subscriber = self.create_subscription(
+            ObjectDetection2dSet,
+            self._obj_det_topic,
+            self.obj_det_callback,
+            1000 # TODO: queue size
+        )
+
+        self._publisher = self.create_publisher(
+            ActivityDetection,
+            self._det_topic,
+            1
+        )
+
+        # Stores the data until we have enough to send to the detector
+        self._frames = []
+        self._hand_poses = dict(
+            lhand=[],
+            rhand=[],
+        )
+        self._frame_stamps = []
+        self._obj_dets = []
+
+        # TODO - parameterize this?
+        self._topk = 5
+
+        # Instantiate the activity detector models
+        fcn = UnifiedFCNModule(net="resnext", num_cpts=21, obj_classes=9, verb_classes=12)
+        temporal = TemTRANSModule(act_classes=27, hidden=256, dropout=0.1, depth=6)
+
+        self._detector: UnifiedHOModule = UnifiedHOModule(
+            fcn=fcn,
+            temporal=temporal,
+            checkpoint=self._model_checkpoint,
+            device=self._torch_device,
+            labels_file=self._labels_file
+        )
+        self._detector.eval()
+        self._detector = self._detector.to(device=self._torch_device)
+        log.info(f"UHO Detector initialized")
+
+    def multimodal_listener_callback(self, image, hand_pose):
+        """
+        Callback function images + hand poses. Messages are synchronized with
+        with the ROS time synchronizer.
+        """
+        # Convert ROS img msg to CV2 image and add it to the frame stack
+        rgb_image = BRIDGE.imgmsg_to_cv2(image, desired_encoding="rgb8")
+        rgb_image_np = np.asarray(rgb_image)
+
+        self._frames.append(rgb_image_np)
+
+        # Store the image timestamp
+        self._frame_stamps.append(image.header.stamp)
+
+        lhand, rhand = self.get_hand_pose_from_msg(hand_pose)
+        self._hand_poses['lhand'].append(lhand)
+        self._hand_poses['rhand'].append(rhand)
+
+    def obj_det_callback(self, msg):
+        """
+        Callback for the object detection set message. Once there is a object
+        detection message for each frame in the current set, the activity detector
+        model is called and a new activity detection message is published with
+        the current activity predictions.
```

Comment:
I think this description is no longer accurate: this method is no longer associating detections to every frame.